pkg/frontend/transport/handler.go

```diff
@@ -187,7 +187,12 @@ func (f *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	// Exclude remote read here as we don't have to buffer its body.
 	if !strings.Contains(r.URL.Path, "api/v1/read") {
 		if err := r.ParseForm(); err != nil {
-			writeError(w, err)
+			if util.IsRequestBodyTooLarge(err) {
+				err = errRequestEntityTooLarge
+			} else {
+				err = httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+			}
+			server.WriteError(w, err)
```

Comment:
We can just use `http.Error(w, err.Error(), http. StatusBadRequest)` for the two cases. We don't have to call `server.WriteError` anymore.